openfl-docker/Dockerfile.base

```diff
@@ -13,28 +13,27 @@ SHELL ["/bin/bash", "-o", "pipefail", "-c"]
 RUN --mount=type=cache,id=apt-dev,target=/var/cache/apt \
     apt-get update && \
     apt-get install -y \
+    git \
     python3-pip \
     python3.10-dev \
     ca-certificates \
     build-essential \
-    git \
     --no-install-recommends && \
+    apt-get purge -y linux-libc-dev && \
     rm -rf /var/lib/apt/lists/*
-RUN apt-get purge -y linux-libc-dev
 
 # Create an unprivileged user.
 RUN groupadd -g 1001 default && \
-    useradd -m -u 1001 -g default openfl
-USER openfl
+    useradd -m -u 1001 -g default user
+USER user
+WORKDIR /home/user
+ENV PATH=/home/user/.local/bin:$PATH
 
 # Install OpenFL.
-WORKDIR /home/openfl
-COPY --chown=openfl:default . .
-ENV PATH=/home/openfl/.local/bin:$PATH
-RUN pip install --no-cache-dir -U pip setuptools wheel && \
-    pip install --no-cache-dir -e .
-
-# Download thirdparty licenses.
-RUN INSTALL_SOURCES=yes /home/openfl/openfl-docker/licenses.sh
+RUN git clone --depth=1 --branch=develop https://github.com/securefederatedai/openfl.git && \
```

Comment:
This might limit if someone wants to build/test on there fork / branch as it will always checkout upstream develop

Suggestion: better to take that as param to base file and supply upstream and develop branch as defaults

Reply:
AFAIU, in the original Dockerfile, the container image was built based on the current folder (supposedly the openfl repo root) via the `COPY --chown=openfl:default . .` instruction.

As @ishaileshpant points out, why do we need to change this approach?

Reply:
Agree. This is editable in OpenFL-Security for the same reason you mention. But I don't see this option being widely used here. I initially tested an editable version, but then retracted for multiple reasons:

1. The added complexity of forwarding URLs (`fx -> build-arg -> docker build`) is not worth the benefit gained.
2. It is trivial to change the URL by simply editing the base Dockerfile.
3. Dockerfile (and the associated `fx` command code) becomes less readable and adds maintenance cost.
4. Users typically edit their workspace source code, and base images are intended to be stable (unless they edit OpenFL source code, which we should not require them to, for it must be stable and bug-free).

That said, if there is a high _user_ consensus on making URLs configurable via `fx` command flags, that can be addressed in a future PR.

Reply:
May be not via fx command flags but atleast in the Dockerfile.base itself they can be exposed as build-args and them defaulted to upstream - will be beneficial to openfl devs and not just users - every code changes that has to be validated in container, in current state will mandate dockerfile change as well instead of supplying the fork/branch to validate in PR as build args - does that make sense?

Reply:
As a compromise between this implementation and the pre-existing one, perhaps we can retain the `--base-image` flag in `fx workspace dockerize` to use a pre-built base image so that the user can build from their local/remote branch if they require or proceed with the latest develop code if the param is not used.

Reply:
I agree with @MasterSkepticista's simplification proposal. However, for the sake of stability, I suggest to get a specific stable tag (such as `v1.6`), instead of `develop`, which could regularly contain breaking changes.

Reply:
> AFAIU, in the original Dockerfile, the container image was built based on the current folder (supposedly the openfl repo root) via the `COPY --chown=openfl:default . .` instruction.
> 
> As @ishaileshpant points out, why do we need to change this approach?

My initial approach was to retain copying the OpenFL source code. I retracted it for following reasons:
1. Ideally we'd want a user to be able to run `fx workspace dockerize` anywhere a workspace is. Building a base image with copied source requires specifying build context. This would require adding another flag in `dockerize` to point to the source code. In other words, intention was to keep `fx workspace dockerize` portable, regardless of where the workspace is located on the file system. What if we don't want to provide this flag? We can do this via `SITEPACKS`.
2. Now `SITEPACKS` value depends on what _kind_ of the build it is. Example: we fetch dockerfile using `SITEPACKS/openfl-docker/Dockerfile.X`. Now two ugly cases emerge:
    - If user had installed OpenFL via `pip install -e .`, then `SITEPACKS` points to the OpenFL root. This heavily bloats the docker build context with venv and workspace data. There are ways around it ofcourse, but again, not worth fighting that complexity.
    - If user had installed OpenFL via `pip install .`, then `SITEPACKS` is `.../lib/python/site-packages`. This will pass _all_ packages into build context of base image, and will fail because it does not contain OpenFL installation code (setup.py etc.).
3. User will anyways have access to internet because build requires installation of pip packages. They can simply point to a fork of their choice and avoid all other debug hassle.

Reply:
For anyone working on core framework features/bugs that needs to be tested / validated in container env will have to change the Dockerfile to their fork/branch to test/validate and in the PR will have to revert that back to the upstream - whereas, imo, keeping it configurable at the Dockerfile.base level will make it more flexible with no such additional overhead of maintenance as defaults can still be develop on upstream